NFFT for space, time and frequency domain

rohandahale opened the pull request:
This pull request adds a functionality to support NFFT for ContinuousImages with either of X,Y ; X,Y,T ; X,Y,F ; X,Y,T,F ; X,Y,F,T space, time and frequency dimensions. This should also work for DFT already. The FT is also auto-differentiable with Enzyme and time complexity of the FT is linear with number of times or frequency stamps.

ptiede commented on test/multidomain.jl:
```diff
@@ -0,0 +1,359 @@
+# Test function to check autodiff
+function check4dautodiff(Nx, Nt, Nf, x, dx)
+    p = domains4d(Nx, Nt, Nf)
+    Enzyme.API.runtimeActivity!(true)
```
You don't need to set runtimeActivity! everytime. This can just be put at the top of the script.

ptiede commented:
This looks great! I've left some comments. One thing is that I'd like to run some additional tests on the NFFT and DFT stuff. 
A really simple test is checking that if you handle the NFFT separately for each time/frequency and the automated one that they give identical answers.

ptiede commented on src/fourierdomain/nuft/nuft.jl:
```diff
@@ -8,21 +8,121 @@ Internal type used to store the cache for a non-uniform Fourier transform (NUFT)
+            push!(iminds, (j, i))
+            push!(visinds, findall(p -> p.Ti == ti && p.Fr == fr, vis_points))
+        end
+    end
+    return (iminds, visinds)
+end
+
+function plan_indices(imgdomain::AbstractRectiGrid{<:Tuple{X,Y,Ti}},
+                      visdomain::UnstructuredDomain)
+    iminds = Int[]
+    visinds = Vector{Int}[]
+    vis_points = domainpoints(visdomain)
+    Ti = imgdomain.Ti
+
+    for (i, ti) in pairs(Ti)
+        push!(iminds, i)
+        push!(visinds, findall(p -> p.Ti == ti, vis_points))
+    end
+    return (iminds, visinds)
+end
+
+function plan_indices(imgdomain::AbstractRectiGrid{<:Tuple{X,Y,Fr}},
+                      visdomain::UnstructuredDomain)
+    iminds = Int[]
+    visinds = Vector{Int}[]
+    vis_points = domainpoints(visdomain)
+    Fr = imgdomain.Fr
+
+    for (j, fr) in pairs(Fr)
+        push!(iminds, j)
+        push!(visinds, findall(p -> p.Fr == fr, vis_points))
+    end
+    return (iminds, visinds)
+end
+
+# For only spatial case
+function plan_indices(imgdomain::AbstractRectiGrid{<:Tuple{X,Y}},
+                      visdomain::UnstructuredDomain)
+    return (0, 0)
+end
```
We can replace this entire block of code with 

```julia
compare(nv::NamedTuple{N}, val) where {N} = mapreduce(n->(nv[n] == val[n]), *, N)
function plan_indices(imgdomain, visdomain)
    spatialdims = ComradeBase.dims(imgdomain)[3:end]
    nms = map(name, spatialdims)

    isempty(spatialdims) && return (0, 0)

    itr = Iterators.product(map(eachindex, spatialdims)...)
    iminds = eltype(itr)[]
    visinds = Vector{Int}[]
    visp = domainpoints(visdomain)
    for (i, vals) in pairs(DimPoints(spatialdims))
        nv = NamedTuple{nms}(vals)
        push!(iminds, i)
        push!(visinds, findall(p ->compare(nv, p), visp))
    end
    return iminds, visinds
end
```

ptiede commented on src/fourierdomain/nuft/nuft.jl:
```diff
@@ -8,21 +8,121 @@ Internal type used to store the cache for a non-uniform Fourier transform (NUFT)
+    return (0, 0)
+end
+
+# Function has been modified to process imgdomain/visdomain with Ti or Fr or both.
+# It calls the new plan_nuft  when Ti or Fr is present or else calls
+# the old spatial function
```
We don't need these modified comments in the code. It will be pretty easy to forget when/why this was modified.

ptiede commented on src/fourierdomain/nuft/nuft.jl:
```diff
@@ -8,21 +8,121 @@ Internal type used to store the cache for a non-uniform Fourier transform (NUFT)
+    return NUFTPlan(algorithm, plan, phases, indices, size(visdomain)[1])
 end
 
+# Added plan indices and totalvis points to the NUFTPlan in this function
```
Again remove the description of what was changed from the code. This can go in the PR, but honestly it is pretty clear from the code what it does.

ptiede commented on src/fourierdomain/nuft/nuft.jl:
```diff
@@ -8,21 +8,121 @@ Internal type used to store the cache for a non-uniform Fourier transform (NUFT)
+    return NUFTPlan(plan.alg, plan.plan', inv.(plan.phases), plan.indices, plan.totalvis)
+end
+
+# This a new function is overloaded to handle when NUFTPlan has plans
+# as dictionaries in the case of Ti or Fr case
```
Again remove these comments

rohandahale commented on src/fourierdomain/nuft/nfft_alg.jl:
```diff
@@ -31,27 +31,86 @@ Base.@kwdef struct NFFTAlg{T,N,F} <: NUFT
+    for i in eachindex(iminds, visinds)
+        imind = iminds[i]
+        visind = visinds[i]
+        vis_inner = nuft(plans[imind], @view(img[:, :, imind...]))
```
I have added a comment that this a To-do.

rohandahale commented on src/fourierdomain/nuft/nfft_alg.jl:
```diff
@@ -31,27 +31,86 @@ Base.@kwdef struct NFFTAlg{T,N,F} <: NUFT
+        imind = iminds[i]
+        visind = visinds[i]
+        vis_inner = nuft(plans[imind], @view(img[:, :, imind...]))
+        vis_list[visind] .= vis_inner
```
Same for this. I have added a comment that this is a To-do.

rohandahale commented:
I have shorted `plan_indices `  as suggested, moved `plan_nuft` to `nuft.jl`, added tests for DFT.  There is also a test for checking if NFFT/DFT done separately for each time/frequency = the automated one.

ptiede reviewed (approved):
Looks good! I think we are ready to merge

ptiede commented on src/fourierdomain/nuft/nfft_alg.jl:
```diff
@@ -33,10 +33,6 @@ end
-# This function should be in ComaradeBase
 function Base.getindex(domain::UnstructuredDomain; Ti=nothing, Fr=nothing)
     points = domainpoints(domain)
     indices = if Ti !== nothing && Fr !== nothing
```
Remove getindex since now in ComradeBase

ptiede commented:
@rohandahale something funky is going on with the tests. Does this still pass locally for you.

rohandahale commented:
@ptiede  DFT was casing memory issue so I reduced the time and frequency stamps and all tests pass now. But for some reason formatter returns true locally but not here.

ptiede commented:
Ok looks good. Merging
